samples/SampleApp/Startup.cs

```diff
@@ -71,11 +94,34 @@ public static Task Main(string[] args)
 
                     options.ListenLocalhost(basePort + 2, listenOptions =>
                     {
-                        listenOptions.UseHttps("testCert.pfx", "testPassword");
+                        // Use default dev cert
+                        listenOptions.UseHttps();
                     });
 
                     options.ListenAnyIP(basePort + 3);
 
+                    options.ListenAnyIP(basePort + 4, listenOptions =>
+                    {
+                        listenOptions.UseHttps(StoreName.My, "aspnet.test", StoreLocation.CurrentUser, allowInvalid: true);
+                    });
+
+                    options
+                        .Configure()
+                        .Endpoint(IPAddress.Loopback, basePort + 5)
+                        .LocalhostEndpoint(basePort + 6)
+                        .Build();
```

Comment:
So, this I'm guessing Build() is required because that's what triggers the endpoint to actually get added to the options?  

Related question, there is no Build() missing on the Kestrel binding one below, so would that just basically be a no-op giving you an unused builder?

I think it might make things simpler if you just add a separate dedicated options for the defaults, i.e. KestralServerDefaultOptions and configure it via options normally.  Thent he ConfigureXyzDefault methods would just be configuring this other option instead of manipulating the builder.   

You'd still need the existing builder logic to be able to create endpoints from config, but its no longer part of the KestralServerOptions.  The server can just get new default options along with the normal options and continue doing the same thing that Build() does.  The only change that might be required is all of the defaults configuration would need to be done earlier, as they would need to be in the service container.

I definitely don't have a ton of context in this code or how everything fits together, but using a dedicated options class for the defaults seems to be an potential alternative path to consider

Reply:
Kestrel calls Build on the last builder at Start. The idea is that the metapackage will assign a default configuration but it won't call Build so that that default can be replaced by user code. If you want to add two configurations you need to call Build on at least the first one.

Reply:
Requiring them to call build sometimes seems a bit awkward, can't you just expose a list that they can control and remove/add as needed instead... options.Builders.Add instead of options.Configure

RE duplication, you should be able to reuse whatever sub options you want in the defaults, why wouldn't you be able to use the ListenOptions/Adapters for the default options as well.  But even if you wanted to have a complete identical surface area, you could just use named options

// Store defaults via "default" name
```
   var defaults = IOptionsMonitor<KestrelServerOptions>.Get("defaults");
   // ConfigureXDefault methods just pass extra name parameter
```

Reply:
In practice I expect them to almost never call build, they either use the config defined in the metapackage or they use the one from their own code. Keeping a list does not seem necessary. We'll see what the preview1 feedback is like.

From your examples I'm not sure you understand what defaults we're talking about configuring. It's not that there are multiple KestrelServerOptions with a set of defaults, it's that a given KestrelServerOptions has multiple endpoints / ListenOptions / httpsOptions and you can define defaults for the options on those. The connection adapters are the most tricky here, you'd need to create a fresh one for each endpoint so you can't create one as a default without implementing some kind of cloning logic. With the current model you can re-use all of the extensions off of ListenOptions for creating connection adapters.

Reply:
So let me try to get on the same page...

So the KestrelServer eventually wants to produce a bunch of endpoints which has options in a ListenOptions instance, (Endpoints are created from these 1:1?) 

Https is doing weird stuff which I guess relates to the fact that its configuring listen option instances and doing its own options like stuff.  The EndpointDefaults action is basically equivalent to doing a ConfigureAll on ListenOptions if you were using IOptions.

Part of the confusion here I think is that KestrelServerOptions are using IOptions/Configure, while there are a few options types that are used more like EndpointSettings/AdapterSettings and not configured like normal options.

With named endpoints, that matches exactly named options, you can just use `IOptions<EndpointOptions>.Get(name)` and just bind against `config.GetSection("Kestrel:Endpoints")`.  Your defaults could then just be sugar for services.ConfigureAll<EndpointOptions>(o =>  o.Protocols = HttpProtocols.Http1)


Assuming you can just hang all Endpoint operations off of `.Endpoints`, maybe something like?   But I'm sure I don't grok what's going on around the adapter/https yet.

```C#
   UseKestrel(context, o => {

    // Default wireup (but showing what the sugar methods call)
    o.ConfigureAll<EndpointOptions>(opt => opt.Protocols = HttpProtocols.Http1);
    o.ConfigureAll<HttpsDefaultOptions>(o => o.SslProtocols = SslProtocols.Tls12);

      // just binds all the endpoint options to their names using Configure (and defaults?)
      o.Endpoints.Configure(context.Configuration.GetSection("Kestrel:Endpoints")); 
      o.Endpoints.Add(IPAddress.Loopback, basePort + 5);
      o.Endpoints.AddLocalhost(basePort + 6);
      o.Endpoints.AddNamed("NamedEndpoint", opt =>
                        {
                             opt.Listener.Protocols = HttpProtocols.Http1;
                         })
      o.Endpoints.AddNamed("NamedHttpsEndpoint", opt =>
                         {
                             opt.Https.SslProtocols = SslProtocols.Tls12;
                         });
   }
```

Reply:
You're still missing quite a bit of context. Stop by if you want to walk through it.

The question I'd originally sent you was if configuration has a static helper for the scenario where I only want an empty placeholder instance like `new ConfigurationBuilder().Build()`

Reply:
No because there's always config, its pretty strange to pass in empty configuration to a configuration loader class though...  Hence why I think the naming is a bit weird.

Like this doesn't use config at all right?
```
options
                         .Configure()
                         .Endpoint(IPAddress.Loopback, basePort + 5)
                         .LocalhostEndpoint(basePort + 6)
                         .Load();
```

Reply:
Damian wanted to mirror the Listen APIs so you could still configure everything in one place. Yes it makes for an slightly odd fraken type.

Reply:
That's fine, my main point is that this seems more than just a configuration loader, this is public endpoint management API basically, where the ctor takes an optional config and Load() is basically committing the set of endpoints.  

I think building that dedicated API where configuration is just one way to import a chunk of endpoint settings is maybe a simpler way to expose this...

I still think this would be easier to understand if the API just exposed the endpoints on the options.  Having to call configure/endpoint/load and then another configure/endpoint block just seems strange.

Seems like it'd be better to let them express it more concisely, without any weird loads.

```
options.AddEndpoint(IPAddress.Loopback, basePort + 5)
    .AddLocalhostEndpoint(basePort + 6)
    .LoadConfig(context.Configuration.GetSection("Kestrel")
    .AddEndpoint("NamedEndpoint", opt =>
                         {
                             opt.ListenOptions.Protocols = HttpProtocols.Http1;
                         })
                         .Endpoint("NamedHttpsEndpoint", opt =>
                         {
                             opt.HttpsOptions.SslProtocols = SslProtocols.Tls12;
                         }
```

And you can always just let them Remove/Update existing endpoints.